server/src/main/java/org/tctalent/server/service/db/impl/CandidateServiceImpl.java

```diff
@@ -405,2392 +405,2410 @@
-    }
+    computeIeltsScore(candidate);
 
-    @Override
-    public Candidate findByIdLoadSavedLists(long candidateId) {
-        return candidateRepository.findByIdLoadSavedLists(candidateId);
-    }
+  }
 
-    @Override
-    public Candidate findByIdLoadUser(long id, Set<Country> sourceCountries) {
-        return candidateRepository.findByIdLoadUser(id, sourceCountries).orElse(null);
+  private void populateExamIntakeData(@NonNull CandidateExam candidateExam,
+      @NonNull Candidate candidate, CandidateIntakeDataUpdate data) {
+    candidateExam.setCandidate(candidate);
+    if (data.getExamType() != null) {
+      candidateExam.setExam(data.getExamType());
     }
-
-    @Transactional(readOnly = true)
-    public void validateContactRequest(User user, BaseCandidateContactRequest request) {
-        // Check email not already taken
-        if (!StringUtils.isBlank(request.getEmail())) {
-            try {
-                User exists = userRepository.findByEmailIgnoreCase(request.getEmail());
-                if (user == null && exists != null || exists != null && !exists.getId().equals(user.getId())) {
-                    throw new UsernameTakenException("email");
-                }
-            } catch (IncorrectResultSizeDataAccessException e) {
-                throw new UsernameTakenException("email");
-            }
-        }
+    if (data.getOtherExam() != null) {
+      candidateExam.setOtherExam(data.getOtherExam());
     }
-
-    private static String countryStr(String country) {
-        return country == null ? "%" : country;
+    if (data.getExamScore() != null) {
+      // If the ExamScore is NoResponse set to null in database.
+      if (data.getExamScore().equals("NoResponse")) {
+        candidateExam.setScore(null);
+      } else {
+        candidateExam.setScore(data.getExamScore());
+      }
     }
-
-    private static String genderStr(Gender gender) {
-        return gender == null ? "%" : gender.toString();
+    if (data.getExamYear() != null) {
+      candidateExam.setYear(data.getExamYear());
     }
-
-    private static List<DataRow> toRows(List<Object[]> objects) {
-        List<DataRow> dataRows = new ArrayList<>(objects.size());
-        for (Object[] row: objects) {
-            String label = row[0] == null ? "undefined" : row[0].toString();
-            DataRow dataRow = new DataRow(label, (BigInteger)row[1]);
-            dataRows.add(dataRow);
-        }
-        return dataRows;
+    if (data.getExamNotes() != null) {
+      candidateExam.setNotes(data.getExamNotes());
     }
+  }
 
-    @Override
-    public List<DataRow> computeBirthYearStats(Gender gender, LocalDate dateFrom, LocalDate dateTo, List<Long> sourceCountryIds) {
-        return toRows(candidateRepository.
-                countByBirthYearOrderByYear(
-                        genderStr(gender), sourceCountryIds, dateFrom, dateTo));
+  private void populateIntakeData(Candidate candidate, CandidateIntakeDataUpdate data,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 99. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 2321-2391 into a separate method.